### ISSUE TITLE
Odd behavioiur of many'

### ISSUE THREAD

**Issue by erikd** [bos/attoparsec#72]
**TL;DR**
I had a parser that was failing in odd ways and fixed it by replacing

```
 P.many' pSomething
```

with

```
P.option [] (P.many1 pSomething)
```

where pSomething is a non-trivial parser.

Q1: Why does the second verion work correctly when the first doesn't?
Q2: Is there any performance penalty for implementing it the second way?

**Long version**

I had a parser (used in a Conduit via Data.Conduit.Attoparsec) taht somewhere around the time 0.12 was released started intermittently throwing a ParseError (defined in D.C.Attoparsec). However, since it was intermittent and didn't always happen in the same place on the same input (which was ok for my application) I wasn't in a rush to debug it.

When I finally decided to invest some time in debugging this it was really hard going. First I managed to make it trigger more frequently by adding a conduit chunker before the `conduitParser`. Normally ByteString data flows though conduits in chunks of about 30. My chunker allowed me to make these chunks much smaller which increased the frequency of the parse error. Once that was happening I spent considerable time digging around on both D.C.Attoparsec as well as attoparsec itself. I got nowhere.

After reading the attoparsec documentation for about the 10th time I decied to remove all the attoparsec combinators that had warnings about them. I remove `peekWord8` and was more careful about `takeWhile` etc. In the end, it was the removing `many'` that fixed it.
